### PR TITLE
Global state edition

### LINKED ISSUE #346: Global state edition
We need an edition in the global state of our application.

### PR DESCRIPTION
**Pull request by jitsedesmet**
This PR aims to close #346.

### CODE REVIEW

**Comment by jitsedesmet** [comment]
this PR introduces some latency when running with `npm run dev` (I think). The latency is acceptable when running in production though.

File: frontend/src/components/routeInjector.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        if (cachedEdition && !curEditionName) {
+            Router.replace({
+                query: { ...Router.query, edition: cachedEdition.name },
+            }).catch(console.log);
+        }
+
+        if (!cachedEdition && !curEditionName && latestEditionName) {
+            Router.replace({
```
Comment (andpuydt) [inline frontend/src/components/routeInjector.tsx:41]: Same as before concerning the Router.

Comment (andpuydt) [inline frontend/src/components/routeInjector.tsx:41]: Same as before concerning the Router.

**Review by andpuydt — CHANGES_REQUESTED** [review]
Looks good, just a few questions. Since this PR will require a lot of changes on the ongoing branches, could you maybe add an overview of those changes? Such as the use of useSwrWithEdition, the path transformer, an explanation for the changed pathnames, etc. Could you also add some tests?

File: frontend/src/api/calls/baseCalls.tsx [hunk 2]
```diff
@@ -70,17 +73,19 @@ export async function getEntitiesWithCache(
         urls.map(async (url) => {
             if (!cache[url] && !fetched.has(url)) {
                 fetched.add(url);
-                cache[url] = await getEntityOnUrl(url);
+                cache[url] = (await getEntityOnUrl(url))!;
```
[the remark this answers is not in the document]
Comment (jitsedesmet) [reply frontend/src/api/calls/baseCalls.tsx:76]: The type of getEntityOnUrl has changed to also be undefined.

File: frontend/src/components/routeGuard.tsx [hunk 3]
```diff
@@ -1,47 +1,31 @@
...
-
-        // A request to the backend will return redirect to the login when the user was not authenticated
-        if (
-            !publicPaths.includes(path) &&
-            userResponse.request.responseURL == ApiPaths.base + ApiPaths.loginRedirect
-        ) {
-            setAuthorized(false);
-            await push({
+        if (!authException && userErrorMsg) {
+            Router.replace({
```
[the remark this answers is not in the document]
Comment (jitsedesmet) [reply frontend/src/components/routeGuard.tsx:21]: I think this triggerd the useEffect constantly but if you want I can verify this.

**Comment by jitsedesmet** [comment]
> Looks good, just a few questions. Since this PR will require a lot of changes on the ongoing branches, could you maybe add an overview of those changes? Such as the use of useSwrWithEdition, the path transformer, an explanation for the changed pathnames, etc. Could you also add some tests?

What kind of tests are you thinking of? (Also not the urgency behind this PR)

File: frontend/src/components/routeGuard.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (jitsedesmet) [reply frontend/src/components/routeGuard.tsx:21]: Hmm, seems like I was wrong. Thanks for pointing this out.

**Comment by jitsedesmet** [comment]
This PR essentially introduces only 2 functions.
First we have  `useSwrWithEdition` this hook takes the same arguments as useSwr but will make sure the call is queried with the current edition id. The back end is written in a way that every search/by-edition expects the same argument edition so this function generalizes just that.

Second we have `useEditionPathTransformer` this hook returns a function that when provided with a front end path, return the same path but with the front end edition query. Whenever you route to a page that is edition dependent, you should use this transformer.

I hope this is clear. If not, please say so.

**Review by andpuydt — APPROVED** [review]

**Comment by andpuydt** [comment]
> This PR essentially introduces only 2 functions. First we have `useSwrWithEdition` this hook takes the same arguments as useSwr but will make sure the call is queried with the current edition id. The back end is written in a way that every search/by-edition expects the same argument edition so this function generalizes just that.
> 
> Second we have `useEditionPathTransformer` this hook returns a function that when provided with a front end path, return the same path but with the front end edition query. Whenever you route to a page that is edition dependent, you should use this transformer.
> 
> I also implemented a `useOwnUser` hook. This hook just does what it says.
> 
> I hope this is clear. If not, please say so.

Can you pin this on discord for @ruvmello and @thvmulle so they are aware of the changes as well?

**Pull request merged**